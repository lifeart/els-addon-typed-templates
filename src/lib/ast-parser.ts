import { preprocess as parse, traverse } from "@glimmer/syntax";

import {
  isSimpleBlockComponentElement,
  tagComponentToBlock
} from "./ast-helpers";

export function getClassMeta(source) {
  const nodes: any = [];
  const comments: any = [];
  try {
    const node = parse(source);

    traverse(node, {
      MustacheStatement(node) {
        //@ts-ignore
        if (!node.isIgnored) {
          nodes.push([node]);
        }
      },
      MustacheCommentStatement(node) {
        if (node.loc) {
          comments.push([node.loc.end.line + 1, node.value]);
        }
      },
      BlockStatement(node) {
        nodes.push([node]);
      },
      ElementModifierStatement(node) {
        nodes.push([node]);
      },
      ElementNode(node) {
        if (isSimpleBlockComponentElement(node)) {
          nodes.push([tagComponentToBlock(node)]);
        }
      },
      SubExpression(node) {
        if (nodes[nodes.length - 1]) {
          nodes[nodes.length - 1].push(node);
        } else {
          console.log("unexpectedSubexpression", node);
        }
      }
    });
  } catch(e) {
      // 
  }

  return {
    nodes, comments
  };
}
