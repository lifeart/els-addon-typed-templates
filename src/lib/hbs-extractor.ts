import { keyForItem } from "./hbs-transform";

export function extractRelationships(items) {
  const componentsForImport: string[] = [];
  const parents = {};
  const scopes = {};
  const klass = {};
  const blockPaths: any = [];
  function addChilds(items, key) {
    items.forEach(item => {
      if (item.type === "MustacheStatement" || item.type === "BlockStatement") {
        parents[key].push(keyForItem(item));
      } else if (item.type === "ElementNode") {
        item.modifiers.forEach(mod => {
          parents[key].push(keyForItem(mod));
        });
      }
      addChilds(item.program ? item.program.body : item.children || [], key);
      if (item.inverse) {
        addChilds(item.inverse.body || [], key);
      }
    });
  }

  items.slice(0).forEach(b => {
    let n = b.slice(0);
    let pointer: any = null;
    while (n.length) {
      let exp = n.shift();
      const key = keyForItem(exp);

      if (!pointer) {
        pointer = key;
        parents[pointer] = [];
        scopes[pointer] = exp.program ? exp.program.blockParams : [];
        addChilds(exp.program ? exp.program.body : exp.children || [], pointer);
        addChilds(exp.inverse ? exp.inverse.body : [], pointer);
      }

      klass[key] = exp;
      klass[keyForItem(exp.path)] = exp.path;
      if (exp.type === "BlockStatement") {
        blockPaths.push(keyForItem(exp.path));
        if (exp.isComponent) {
          componentsForImport.push(exp.path.original);
        }
      }
      parents[pointer].push(keyForItem(exp.path));
      exp.params.forEach(p => {
        klass[keyForItem(p)] = p;
        parents[pointer].push(keyForItem(p));
      });
      exp.hash.pairs.forEach(p => {
        klass[keyForItem(p.value)] = p.value;
        parents[pointer].push(keyForItem(p.value));
      });
    }
  });

  return {
    componentsForImport,
    parents,
    scopes,
    klass,
    blockPaths
  };
}
