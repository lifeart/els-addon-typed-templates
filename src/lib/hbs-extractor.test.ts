import { getClassMeta } from './ast-parser';
import { extractRelationships } from './hbs-extractor';

function results(source) {
    let items = extractRelationships(getClassMeta(source).nodes);
    delete items.klass;
    return items;
}

describe('getClassMeta', () => {
    const cases = [
        '{{1}}',
        `{{"1"}}`,
        `{{foo}}`,
        `{{#let foo as |bar|}}{{bar}}{{/let}}`,
        '<MyComponent as |ctx|> {{ctx.result}} </MyComponent>',
        `{{foo this.bar baz=this.boo}}`,
        '<input {{foo-bar this.baz a="1"}}>',
        '<MyComponent as |ctx|>{{#let ctx.result as |bar|}}{{bar}}{{/let}}</MyComponent>',
    ];
    cases.forEach((str)=>{
        it(str, ()=>{
            expect(results(str)).toMatchSnapshot();
        });
    });
});