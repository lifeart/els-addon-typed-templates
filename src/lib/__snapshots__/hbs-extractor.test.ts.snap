// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getClassMeta <MyComponent as |ctx|> {{ctx.result}} </MyComponent> 1`] = `
Object {
  "blockPaths": Array [
    "1,0:1,52 - PathExpression",
  ],
  "componentsForImport": Array [
    "my-component",
  ],
  "parents": Object {
    "1,0:1,52 - BlockStatement": Array [
      "1,23:1,37 - MustacheStatement",
      "1,0:1,52 - PathExpression",
    ],
    "1,23:1,37 - MustacheStatement": Array [
      "1,25:1,35 - PathExpression",
    ],
  },
  "scopes": Object {
    "1,0:1,52 - BlockStatement": Array [
      "ctx",
    ],
    "1,23:1,37 - MustacheStatement": Array [],
  },
}
`;

exports[`getClassMeta <MyComponent as |ctx|>{{#let ctx.result as |bar|}}{{bar}}{{/let}}</MyComponent> 1`] = `
Object {
  "blockPaths": Array [
    "1,0:1,79 - PathExpression",
    "1,25:1,28 - PathExpression",
  ],
  "componentsForImport": Array [
    "my-component",
  ],
  "parents": Object {
    "1,0:1,79 - BlockStatement": Array [
      "1,22:1,65 - BlockStatement",
      "1,50:1,57 - MustacheStatement",
      "1,0:1,79 - PathExpression",
    ],
    "1,22:1,65 - BlockStatement": Array [
      "1,50:1,57 - MustacheStatement",
      "1,25:1,28 - PathExpression",
      "1,29:1,39 - PathExpression",
    ],
    "1,50:1,57 - MustacheStatement": Array [
      "1,52:1,55 - PathExpression",
    ],
  },
  "scopes": Object {
    "1,0:1,79 - BlockStatement": Array [
      "ctx",
    ],
    "1,22:1,65 - BlockStatement": Array [
      "bar",
    ],
    "1,50:1,57 - MustacheStatement": Array [],
  },
}
`;

exports[`getClassMeta <input {{foo-bar this.baz a="1"}}> 1`] = `
Object {
  "blockPaths": Array [],
  "componentsForImport": Array [],
  "parents": Object {
    "1,7:1,33 - ElementModifierStatement": Array [
      "1,9:1,16 - PathExpression",
      "1,17:1,25 - PathExpression",
      "1,28:1,31 - StringLiteral",
    ],
  },
  "scopes": Object {
    "1,7:1,33 - ElementModifierStatement": Array [],
  },
}
`;

exports[`getClassMeta {{"1"}} 1`] = `
Object {
  "blockPaths": Array [],
  "componentsForImport": Array [],
  "parents": Object {
    "1,0:1,7 - MustacheStatement": Array [
      "1,2:1,5 - StringLiteral",
    ],
  },
  "scopes": Object {
    "1,0:1,7 - MustacheStatement": Array [],
  },
}
`;

exports[`getClassMeta {{#let foo as |bar|}}{{bar}}{{/let}} 1`] = `
Object {
  "blockPaths": Array [
    "1,3:1,6 - PathExpression",
  ],
  "componentsForImport": Array [],
  "parents": Object {
    "1,0:1,36 - BlockStatement": Array [
      "1,21:1,28 - MustacheStatement",
      "1,3:1,6 - PathExpression",
      "1,7:1,10 - PathExpression",
    ],
    "1,21:1,28 - MustacheStatement": Array [
      "1,23:1,26 - PathExpression",
    ],
  },
  "scopes": Object {
    "1,0:1,36 - BlockStatement": Array [
      "bar",
    ],
    "1,21:1,28 - MustacheStatement": Array [],
  },
}
`;

exports[`getClassMeta {{1}} 1`] = `
Object {
  "blockPaths": Array [],
  "componentsForImport": Array [],
  "parents": Object {
    "1,0:1,5 - MustacheStatement": Array [
      "1,2:1,3 - NumberLiteral",
    ],
  },
  "scopes": Object {
    "1,0:1,5 - MustacheStatement": Array [],
  },
}
`;

exports[`getClassMeta {{foo this.bar baz=this.boo}} 1`] = `
Object {
  "blockPaths": Array [],
  "componentsForImport": Array [],
  "parents": Object {
    "1,0:1,29 - MustacheStatement": Array [
      "1,2:1,5 - PathExpression",
      "1,6:1,14 - PathExpression",
      "1,19:1,27 - PathExpression",
    ],
  },
  "scopes": Object {
    "1,0:1,29 - MustacheStatement": Array [],
  },
}
`;

exports[`getClassMeta {{foo}} 1`] = `
Object {
  "blockPaths": Array [],
  "componentsForImport": Array [],
  "parents": Object {
    "1,0:1,7 - MustacheStatement": Array [
      "1,2:1,5 - PathExpression",
    ],
  },
  "scopes": Object {
    "1,0:1,7 - MustacheStatement": Array [],
  },
}
`;
